### ISSUE TITLE
PyInstaller doesn't respect LSUIElement=1 in OS X

### ISSUE THREAD

**Issue by micahflee** [pyinstaller/pyinstaller#1917]
I'm making a python3/PyQt5 application for Mac OS X that's supposed to only be an applet icon, and should not have an icon on the dock or a menu bar. The way to do this is to set LSUIElement to 1 in the `Info.plist` file ([see here](https://developer.apple.com/library/ios/documentation/General/Reference/InfoPlistKeyReference/Articles/LaunchServicesKeys.html#//apple_ref/doc/uid/20001431-108256) for more info). I can specify custom `Info.plist` keys and values [easily](https://pythonhosted.org/PyInstaller/#spec-file-options-for-a-mac-os-x-bundle), and I can confirm that it successfully gets in my `Info.plist` after building a .app with PyInstaller.

But when I run my application it does not act like it has LSUIElement set -- there is an icon in the dock, the application has a menu bar, and it shows up when I press Apple-Tab and in Force Quit.

I'm not sure, but after a lot of searching and troubleshooting I believe that this is related to how PyInstaller bootstraps the application. Specifically, in `bootloader/src/pyi_launch.c` the call to `TransformProcessType` in the [pyi_launch_initialize function](https://github.com/pyinstaller/pyinstaller/blob/develop/bootloader/src/pyi_launch.c#L482) undoes LSUIElement.

``` c
void
pyi_launch_initialize(ARCHIVE_STATUS * status)
{
#if defined(__APPLE__) && defined(WINDOWED)
    /*
     * On OS X this ensures that the application is handled as GUI app.
     * Call TransformProcessType() in the child process.
     */
    ProcessSerialNumber psn = { 0, kCurrentProcess };
    OSStatus returnCode = TransformProcessType(&psn,
                                               kProcessTransformToForegroundApplication);
#elif defined(_WIN32)
    char * manifest;
    manifest = pyi_arch_get_option(status, "pyi-windows-manifest-filename");

    if (NULL != manifest) {
        manifest = pyi_path_join(NULL, status->mainpath, manifest);
        CreateActContext(manifest);
        free(manifest);
    }
#endif /* if defined(__APPLE__) && defined(WINDOWED) */
}
```

When the application bootstraps, the parent pyinstaller process becomes a background application and the child application (the actual python app) becomes a foreground application -- even if its `Info.plist` file specifies that it's supposed to be a background application.

I'm not entirely clear the exact relationship between LSUIElement and LSBackgroundOnly, but I think that if the application's `Info.plist` file specifically sets either of those to true, then the `TransformProcessType` call should actually be passed `kProcessTransformToBackgroundApplication` rather than `kProcessTransformToForegroundApplication`.

**Comment by micahflee** [comment]
I'm troubleshooting this more, and it looks like it might not be `TransformProcessType`'s fault. I've tried modifying the bootloader to check for an environment variable called `_LSUIELEMENT`, if it is set to "1" then passing `kProcessTransformToBackgroundApplication` into `TransformProcessType`, and otherwise passing `kProcessTransformToForegroundApplication`.

But it doesn't make a different, both ways still make an icon appear in the dock. I've also tried just not calling `TransformProcessType` at all if the environment variable is set, but that doesn't help either. I'm sure that my environment variable setting is being correctly read.

Here's how I changed that function:

```c
void
pyi_launch_initialize(ARCHIVE_STATUS * status)
{
#if defined(__APPLE__) && defined(WINDOWED)
    /*
     * On OS X this ensures that the application is handled as GUI app.
     * Call TransformProcessType() in the child process.
     */

    ProcessSerialNumber psn = { 0, kCurrentProcess };
    OSStatus returnCode;

    char *lsuielement = NULL;
    lsuielement = pyi_getenv("_LSUIELEMENT");
    if(lsuielement != NULL && strcmp(lsuielement, "1") == 0) {
        /* If the environment variable _LSUIELEMENT is set to "1", then make
         * sure the child process stays in the background.
         */
        printf("_LSUIELEMENT is set, making background app");
        returnCode = TransformProcessType(&psn, kProcessTransformToBackgroundApplication);
    } else {
        /* If _LSUIELEMENT is not set to "1", then make sure the child process
         * becomes a foreground application.
         */
         printf("_LSUIELEMENT is not set, making foreground app");
         returnCode = TransformProcessType(&psn, kProcessTransformToForegroundApplication);
    }
```

Does anyone who is more familiar with the inner working of PyInstaller have an ideas?

**Comment by Tristan79** [comment]
Same here,... 

https://github.com/beville/ComicStreamer/blob/master/mac/build_notes.txt

solved it but, which line to patch?

**Comment by Tristan79** [comment]
I patched, pyi_launch.c, removed the 1 line... recompile it with waf all, 
moved manually the 4 build files to /PyInstaller/bootloader/Darwin-64bit
and run setup.py build and install... 

now I do not have a icon anymore!!!! ;-)

round line 480....   somebody should build real patch here, check, is the option in the plist is set if its an agent...  


```
/* For finer grained control. */

void
pyi_launch_initialize(ARCHIVE_STATUS * status)
{
#if defined(__APPLE__) && defined(WINDOWED)
    /*
     * On OS X this ensures that the application is handled as GUI app.
     * Call TransformProcessType() in the child process.
     */
    ProcessSerialNumber psn = { 0, kCurrentProcess };
    //OSStatus returnCode = TransformProcessType(&psn,
                                             //  kProcessTransformToForegroundApplication);
#elif defined(_WIN32)
    char * manifest;
    manifest = pyi_arch_get_option(status, "pyi-windows-manifest-filename");

    if (NULL != manifest) {
        manifest = pyi_path_join(NULL, status->mainpath, manifest);
        CreateActContext(manifest);
        free(manifest);
    }
#endif /* if defined(__APPLE__) && defined(WINDOWED) */
}

/*
```

**Comment by cbenhagen** [comment]
@Tristan79 are you sure this is all you changed? I am unable to reproduce your results. Would love to see a working pull request.

**Comment by Tristan79** [comment]
Yes. I got it working perfectly for this project. https://github.com/Tristan79/iBrew/. 
I also forked pyinstaller (such that I do not have to do this again: https://github.com/Tristan79/pyinstaller) which includes the recompiled bootloaders and the patch.

**Comment by Tristan79** [comment]
Update, ibrew is based on code from comicstreamer... (https://github.com/Tristan79/ComicStreamer) with ibrew it works... with comicstreamer it hides the icon, but comes back again... so anybody got a proper fix already?

**Comment by Tristan79** [comment]
It does work, i spend a couple of hours scratching my head my it didn't work... on comicstreamer and it does on ibrew, now days later I realized i forgot to add the right plist...

I forgot you also have to modify it and add

https://github.com/Tristan79/iBrew/blob/master/source/distro/mac/Info.plist#L5
https://github.com/Tristan79/iBrew/blob/master/source/distro/mac/Info.plist#L23

so now it also works on comicstreamer :-)

No icon in dock :-)

**Comment by Tristan79** [comment]
It does work, i spend a couple of hours scratching my head my it didn't work... on comicstreamer and it does on ibrew, now days later I realized i forgot to add the right plist...

I forgot you also have to modify it and add

https://github.com/Tristan79/iBrew/blob/master/source/distro/mac/Info.plist#L5
https://github.com/Tristan79/iBrew/blob/master/source/distro/mac/Info.plist#L23

so now it also works on comicstreamer :-)

No icon in dock :-)

**Comment by tnarik** [comment]
Is this related somehow to https://github.com/pyinstaller/pyinstaller/pull/1206?
It seems this modification is not part of any milestone/current development, which is a pity.

**Comment by tnarik** [comment]
Switching from `kProcessTransformToForegroundApplication` to `kProcessTransformToUIElementApplication` in `osx.py` (as hinted by #1206) seems to do the trick.
That is what I just started using today in my local clone of `pyinstaller`. I'm not familiar with the rest of the code, but will take a look to see if there is a way to update the logic so that all cases are supported on macOS nicely. If so, I'll produce a PR.

**Comment by htgoebel** [comment]
@tnarik Would you please be so kind submitting a pull-request for this. Thanks!

Please see our [Guidelines for Commits](https://pyinstaller.readthedocs.io/en/latest/development/commit-messages.html). Our [Development Guide](https://pyinstaller.readthedocs.io/en/latest/development/) might be useful, too.

**Comment by BoboTiG** [comment]
@htgoebel and guys, I am working on. This is a major issue for our [product](https://github.com/nuxeo/nuxeo-drive). Expect a fix in a week or less.

**Comment by htgoebel** [comment]
I expect this to be fixed by merging #3566.
